### PR TITLE
ARROW-5351: [Rust] Take kernel

### PR DESCRIPTION
**Pull request by nevi-me**
https://issues.apache.org/jira/browse/ARROW-5351

Implement `take` kernel, initial draft that hasn't been benchmarked, and is expected to be rough around the edges.

### CODE REVIEW

**Comment by nevi-me** [comment]
@sunchao I've used ArrayData comparison to test, I had expected this test to fail because of this.

I use Windows, so it looks like ArrayData sometimes fails on Windows when comparing 2 similar arrays by their data. The reason being that null/invalid array slots return non-deterministic data.

I'll attach an example tomorrow when I continue with this PR.

I'm mentioning this because you suggested that we avoid using string/debug for comparison of arrays

**Comment by sunchao** [comment]
Yes, more efforts are required to improve equality check for `ArrayData`, `Array` and others. Currently they just use the derived equality which is not good. We should find a way to ignore the null/invalid data when doing the equality check.

**Comment by nevi-me** [comment]
@andygrove @sunchao @paddyhoran I have a few decisions to make, and would appreciate some guidance/opinions.

I'm mostly done with this, but need to optimise bounds checking and implement options (such as whether to even check for bounds).

*Question:* pandas, numpy, and the C++ version (cc @bkietz) have the option of supplying negative indices. Is this something that we'd want in Rust? If so, I'll change the function's `index` parameter  to `Int32Array` from the current unsigned one.

**Comment by sunchao** [comment]
What does a negative index mean? can you show some code examples where this is applied?

**Comment by nevi-me** [comment]
@sunchao here's a good brief reference from Wes https://issues.apache.org/jira/browse/ARROW-772?focusedCommentId=16340235&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16340235

**Comment by bkietz** [comment]
It's worth noting that out of bounds and negative indices are currently just an error in C++

**Comment by nevi-me** [comment]
Thanks @bkietz, I noticed that there's open JIRAs for them. I wanted to get them out of the way earlier so we can stabilise the `take` signature.

**Comment by nevi-me** [comment]
I've done all I can here.

I noticed that if I have an array with 6 values, `array.is_null(6)` and `array.is_null(7)` won't panic because the array length's within a multiple of 8; so the 7th and 8th values are returned as null in the buffer. I don't know if this is an issue, and if we should document this behaviour.

**Comment by wesm** [comment]
@sunchao can you review? It seems then #4331 can be finished after this is merged

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: Can this be merged with `take_numeric`?

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: `take_test_numeric` -> `test_take_numeric`

File: rust/arrow/src/builder.rs
Comment (sunchao) [inline rust/arrow/src/builder.rs]: It seems the `BinaryBuilder` is poorly named - the original `append_value` should be called something else since it only append a single byte to the binary array. This instead should better be called `append_value`.

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: Should we call this `values`? `array` is self explanatory given that the type is `ArrayRef`.

File: rust/arrow/src/compute/util.rs
Comment (sunchao) [inline rust/arrow/src/compute/util.rs]: - `pub(crate` could be changed to `pub(super)` if this is only used within the `compute` module? or, is it better to put this in `take.rs`?
- the name is a little bit misleading - it gives the impression that this is taking indices from a list - maybe rename to `take_value_indices_from_list` or something else?

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: Some optimizations can be done here such as if it is taking the whole list or a contiguous sublist. This can be left as a TODO though.

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: Is this correct? seems it is using value array's null count.

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: nit: might be better to use `ArrayDataBuilder` here.

**Review by sunchao — COMMENTED** [review]
Thanks @nevi-me for the update and sorry for the delay on reviewing!

File: rust/arrow/src/builder.rs
Comment (nevi-me) [reply rust/arrow/src/builder.rs]: I agree on this, and it gets a bit confusing because I would think that appending a "value" would be more common than appending bytes one-by-one. How about the more direct `append_byte`?

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: Do you mean checking if we have 0 nulls and bypassing the null checks?

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: `fn value(&self, i: usize)` is implemented separately for `ArrowNumericType` bound and `BooleanType`, so I (think) I can't create a common `take_primitive`. If I say `... where T: ArrowPrimitiveType`, I fall short because I need the `fn value()`

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: I still prefer `ArrayData::new()` as it's less verbose and reduces my chances of missing something with the builder, but I'll change to `ArrayDataBuilder`

File: rust/arrow/src/compute/util.rs
Comment (nevi-me) [reply rust/arrow/src/compute/util.rs]: Can we keep it here for now? I wanted to use it in a `sort` kernel, which would sort an array and return indices; then use this function (for lists) to take the appropriate value indices.

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [reply rust/arrow/src/compute/kernels/take.rs]: Yes exactly. Sure it can be done in a follow-up.

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [reply rust/arrow/src/compute/kernels/take.rs]: Ideally `ArrayDataBuilder` should do a validation to make sure itself is well-formed before returning the constructed `ArrayData`. The advantage with `ArrayDataBuilder` is that 1): you don't need to handle default values by passing `None`; 2) it is a little bit clearer on the argument-parameter mapping (e.g., which buffer is used as null bit map, which buffer is value buffer, etc).

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [reply rust/arrow/src/compute/kernels/take.rs]: I see. I encountered the exact same issue in another PR. The approach I took is to implement an internal trait just for the `value()` method and have both `PrimitiveArray` and `BooleanArray` implementing it. You can either do it here or we can improve this in a follow-up.

File: rust/arrow/src/builder.rs
Comment (sunchao) [reply rust/arrow/src/builder.rs]: Sure, `append_byte` sounds better.

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: Just a habit, I've become accustomed to importing enums inside the scope that I need them, to avoid repeating `TimeUnit::A`, `TimeUnit::B`, etc. Would you prefer I move it to the top-level?

Comment (sunchao) [reply rust/arrow/src/compute/kernels/take.rs]: I'm not sure whether there's a "official style guide" on this but it may be better to put it in top-level to 1) make it easier to find out what are the dependencies for a module, 2) potentially avoid repeating the use clause in all the places that need it.

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: I'm looking at it now, I can address it in a follow-up as there are other places that can benefit from this dedup.

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: Hi @sunchao, this was quite difficult to deal with; but I've fixed it. The index's null count was only correct if the `values` array had no null counts. I now recompute the null count from the offsets, where an offset of `vec![0,1,1,2,2]` contains 2 null values as 1 and 2 repeat. I'm getting ready to push the latest changes.

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: I ended up using the internal trait that you created, and have merged numeric with boolean into primitive.

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: Done

**Comment by nevi-me** [comment]
Hi @sunchao, PTAL. I've used `ArrayEqual` for my comparisons, correctly calculated list offsets, and fixed other review comments. Apologies for taking very long on this, haven't had capacity in past weeks. I'm available now again, so I'll look at other Rust work.

@liurenjie1024 @andygrove @paddyhoran may you please also review when you can. Thanks

File: rust/arrow/src/compute/kernels/take.rs
Comment (sunchao) [inline rust/arrow/src/compute/kernels/take.rs]: nit: let's fix the format: `TODO Some` -> `TODO: some`. Also break this long line into two.

File: rust/arrow/src/compute/util.rs
Comment (sunchao) [inline rust/arrow/src/compute/util.rs]: nit: add a `:` after `TODO`.

File: rust/arrow/src/compute/util.rs
Comment (sunchao) [inline rust/arrow/src/compute/util.rs]: nit: we can check this right after line 75, on whether `start = end`, so that we don't need to do the extra `map` operation.

File: rust/arrow/src/compute/util.rs
Comment (sunchao) [inline rust/arrow/src/compute/util.rs]: nit: remove this line?

**Review by sunchao — APPROVED** [review]
Thanks @nevi-me ! This looks good. I just have a few more nits - after that we can get this committed.

File: rust/arrow/src/compute/kernels/take.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/take.rs]: I've added an example, but not as a runnable doc example as the function is private. PTAL

**Comment by nevi-me** [comment]
Thanks @sunchao, I've addressed your feedback.

**Comment by sunchao** [comment]
Looks good @nevi-me ! could you try to commit this yourself as you're already a committer now? it's also good for you to learn about the workflow :)

**Comment by nevi-me** [comment]
Thanks @sunchao, I'm still setting up machine for the process, I'll commit it 😃🙏🏾

**Pull request closed without merging**
